File: waltid-applications/waltid-wallet-demo-ios/iosApp/iosApp/Display/CredentialSystemInfo.swift
import Foundation

extension CredentialDetails {
    var systemInfoGroup: ClaimGroup? {
        let items = [
            addedAt.map(Self.systemInfoDateFormatter.string(from:)).systemInfoItem(path: "system.added", label: "Added"),
            id.systemInfoItem(path: "system.id", label: "Credential ID"),
            format.systemInfoItem(path: "system.format", label: "Format"),
            issuer.systemInfoItem(path: "system.issuer", label: "Issuer"),
            subject.systemInfoItem(path: "system.subject", label: "Subject")
        ].compactMap { $0 }

        return items.isEmpty ? nil : ClaimGroup(title: "About this credential", items: items)
    }

    private static let systemInfoDateFormatter: ISO8601DateFormatter = {
        let formatter = ISO8601DateFormatter()
        formatter.timeZone = TimeZone(secondsFromGMT: 0)
        formatter.formatOptions = [.withInternetDateTime]
        return formatter
    }()
}

private extension Optional where Wrapped == String {
    func systemInfoItem(path: String, label: String) -> ClaimItem? {
        self?.systemInfoItem(path: path, label: label)
    }
}

private extension String {
    func systemInfoItem(path: String, label: String) -> ClaimItem? {
        let value = trimmingCharacters(in: .whitespacesAndNewlines)
        guard !value.isEmpty else { return nil }

        return ClaimItem(
            path: ClaimItemPath.topLevel(path),
            label: label,
            value: .text(value),
            rawValue: value
        )
    }
}

File: waltid-applications/waltid-wallet-demo-compose/sharedLogic/src/commonMain/kotlin/id/walt/walletdemo/compose/logic/CredentialSystemInfo.kt
package id.walt.walletdemo.compose.logic

fun CredentialDetails.toSystemInfoGroup(): ClaimGroup? {
    val items = listOfNotNull(
        summary.addedAt.toSystemInfoItem(path = "system.added", label = "Added"),
        summary.id.toSystemInfoItem(path = "system.id", label = "Credential ID"),
        summary.format.toSystemInfoItem(path = "system.format", label = "Format"),
        summary.issuer.toSystemInfoItem(path = "system.issuer", label = "Issuer"),
        summary.subject.toSystemInfoItem(path = "system.subject", label = "Subject"),
    )

    return items.takeIf { it.isNotEmpty() }?.let {
        ClaimGroup(title = "About this credential", items = it)
    }
}

private fun String?.toSystemInfoItem(path: String, label: String): ClaimItem? {
    val value = this?.trim()?.takeIf { it.isNotEmpty() } ?: return null
    return ClaimItem(
        path = ClaimItemPath.topLevel(path),
        label = label,
        value = DisplayValue.Text(value),
        rawValue = value,
    )
}
